File: fs/before_energy/52__n-queens-ii/cpp/n-queens-ii_benchmark_v2.cpp
#include <iostream>
template <class T>
static inline void DoNotOptimize(const T& value) {
    asm volatile("" : : "r,m"(value) : "memory");
}

#include <vector>

class Solution {
public:
    // Time:  O(n!)
    // Space: O(n)
    int totalNQueens(int n) {
        if (n <= 0) return 0;
        std::vector<int> cols(n, 0);
        std::vector<int> main_diag(2 * n - 1, 0);
        std::vector<int> anti_diag(2 * n - 1, 0);
        return dfs(0, n, cols, main_diag, anti_diag);
    }

private:
    int dfs(int row, int n, std::vector<int>& cols,
            std::vector<int>& main_diag, std::vector<int>& anti_diag) {
        if (row == n) {
            return 1;
        }
        int result = 0;
        for (int i = 0; i < n; ++i) {
            int md = row + i;
            int ad = row - i + (n - 1);
            if (cols[i] || main_diag[md] || anti_diag[ad]) {
                continue;
            }
            cols[i] = main_diag[md] = anti_diag[ad] = 1;
            result += dfs(row + 1, n, cols, main_diag, anti_diag);
            cols[i] = main_diag[md] = anti_diag[ad] = 0;
        }
        return result;
    }
};

int main() {
    Solution sol;
    std::vector<int> tests = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};

    volatile int sink = 0;
    const int iterations = 1000;
    for (int iter = 0; iter < iterations; ++iter) {
        //int checksum = 0;
        for (int n : tests) {
            int r = sol.totalNQueens(n);
            DoNotOptimize(r);
        }
        //sink = checksum;
    }

    return 0;
}

File: fs/py_before_energy/52__n-queens-ii/n-queens-ii_benchmark_v2.py
# Time:  O(n!)
# Space: O(n)

class Solution(object):
    def totalNQueens(self, n):
        """
        :type n: int
        :rtype: int
        """
        def dfs(row):
            if row == n:
                return 1
            result = 0
            for i in xrange(n):
                if cols[i] or main_diag[row+i] or anti_diag[row-i+(n-1)]:
                    continue
                cols[i] = main_diag[row+i] = anti_diag[row-i+(n-1)] = True
                result += dfs(row+1)
                cols[i] = main_diag[row+i] = anti_diag[row-i+(n-1)] = False
            return result

        result = []
        cols, main_diag, anti_diag = [False]*n, [False]*(2*n-1), [False]*(2*n-1)
        return dfs(0)

if __name__ == "__main__":
    sol = Solution()
    tests = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

    sink = 0
    iterations = 1000
    for iter in xrange(iterations):
        #checksum = 0
        for n in tests:
            r = sol.totalNQueens(n)
        #sink = checksum